onnxruntime/core/session/inference_session.cc

```diff
@@ -97,29 +98,43 @@ inline std::basic_string<T> GetCurrentTimeString() {
 
 std::atomic<uint32_t> InferenceSession::global_session_id_{1};
 
-InferenceSession::InferenceSession(const SessionOptions& session_options,
-                                   logging::LoggingManager* logging_manager)
-    : session_options_(session_options),
-      graph_transformation_mgr_(session_options.max_num_graph_transformation_steps),
-      logging_manager_(logging_manager),
-      thread_pool_(concurrency::CreateThreadPool("intra_op_thread_pool",
-                                                 session_options.intra_op_num_threads)),
-      inter_op_thread_pool_(session_options.execution_mode == ExecutionMode::ORT_PARALLEL
-                                ? concurrency::CreateThreadPool("inter_op_thread_pool",
-                                                                session_options.inter_op_num_threads)
-                                : nullptr),
-      session_state_(execution_providers_,
-                     session_options.enable_mem_pattern && session_options.execution_mode == ExecutionMode::ORT_SEQUENTIAL,
-                     thread_pool_.get(),
-                     inter_op_thread_pool_.get()),
-      insert_cast_transformer_("CastFloat16Transformer") {
+Status InferenceSession::FinalizeSessionOptions(SessionOptions& session_options,
+                                                const ONNX_NAMESPACE::ModelProto& model_proto) {
+  InferenceSessionUtils inference_session_utils(logging::LoggingManager ::DefaultLogger());
+
+  ORT_RETURN_IF_ERROR(inference_session_utils.ParseOrtConfigJsonInModelProto(model_proto));
+
+  ORT_RETURN_IF_ERROR(inference_session_utils.ParseSessionOptionsFromModelProto(session_options));
+
+  return Status::OK();
+}
+
+void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
+                                         logging::LoggingManager* logging_manager) {
   ORT_ENFORCE(Environment::IsInitialized(),
               "Environment must be initialized before creating an InferenceSession.");
+  session_options_ = session_options;
+  graph_transformation_mgr_ = onnxruntime::make_unique<GraphTransformerManager>(
+      session_options.max_num_graph_transformation_steps);
+  logging_manager_ = logging_manager;
```

Comment:
Instead of making this heap allocated could we just allow changing the number of steps?

Reply:
Resolving this comment per offline discussion.